Optimize Rendering Push Performance Using MIT-SHM

lindexi opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
This pull request implements the use of MIT-SHM for pushing rendered images. It replaces the `XPutImage` method with `XShmPutImage`. This change significantly reduces Bitmap copying at the X11 layer, enhancing rendering push performance.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Currently, only `XPutImage` is used for pushing rendered images in soft rendering mode.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
This pull request provides additional options, allowing developers to configure the use of MIT-SHM shared memory for pushing rendered images. This can effectively improve rendering performance and reduce rendering latency on low-performance devices.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
The solution involves using MIT-SHM to push rendered images, following the method outlined by @kekekeks in https://github.com/AvaloniaUI/Avalonia/discussions/16690#discussioncomment-10378881 .

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

lindexi committed:
Begin define some of the type to X11 XShm

https://github.com/AvaloniaUI/Avalonia/discussions/16690

lindexi committed:
Add libc define

lindexi committed:
Try fix comment.

lindexi committed:
Copy XShm define code

lindexi committed:
Update the comment in XShm.cs

lindexi committed:
Try define the IFramebufferRenderTarget but fail.

lindexi committed:
Add the XShmCompletionEvent

lindexi committed:
Try to get the XShmCompletionEvent without DeferredDisplayEvents

lindexi committed:
Try open the softrender mode tode debug the FramebufferRenderTarget

lindexi committed:
Receive the completion for xshm

lindexi committed:
Check the xshm support before add

lindexi committed:
Pass the visual and depth argument

lindexi committed:
Remove X11ShmImageManager from X11ShmFramebufferContext

lindexi committed:
Finish create X11ShmImage.

lindexi committed:
Fix lost ShmSeg

lindexi committed:
Remove the presentationQueue in X11ShmImageManager

lindexi committed:
Add the PresentationCount

lindexi committed:
If the depth is not 32, we should do some conversion to make the XShmPutImage work. But the conversion is slowly, so we should not use XShmPutImage when the depth is not 32

lindexi committed:
Finish send render

lindexi committed:
Finish dispose the xshm image

lindexi committed:
Move types into matching files

lindexi committed:
Revert "Try open the softrender mode tode debug the FramebufferRenderTarget"

This reverts commit 9e245be7f49c0ff2c11497d5e709598cca033b26.

lindexi committed:
Try to use xshm in linux

lindexi committed:
Add xshm log

lindexi committed:
Dispose X11ShmImage AvailableQueue when X11ShmImageManager Disposed

lindexi committed:
Send render by XShmPutImage

lindexi committed:
Try to output all event

lindexi committed:
Fix using display without lock

lindexi committed:
Force use XSHM

lindexi committed:
Revert "Force use XSHM"

This reverts commit 97e36737f638ae638d634fd055d109d035cd1754.

lindexi committed:
Using `_x11.Display` instead of `_x11.DeferredDisplay`

lindexi committed:
Remove the Render property in X11ShmFramebufferContext

lindexi committed:
Disable the log of Xshm and remove the console message

kekekeks commented on src/Avalonia.X11/X11Structs.cs:
```diff
@@ -772,7 +773,8 @@ internal enum XEventName {
 		MappingNotify		= 34,
         GenericEvent = 35,
-		LASTEvent
+        XShmCompletionEvent = 65,
```
It's not a constant, you are supposed to call `XShmGetEventBase`

lindexi replied:
Thank you @kekekeks .

From https://www.[x.org/releases/X11R7.5/doc/Xext/mit-shm.html](https://www.x.org/releases/X11R7.5/doc/Xext/mit-shm.html)

The XShmCompletionEvent  event type value that will be used can be determined at run time with a line of the form:

    int CompletionType = XShmGetEventBase (display) + ShmCompletion;

kekekeks commented on src/Avalonia.X11/X11Window.cs:
```diff
@@ -195,6 +197,42 @@ public X11Window(AvaloniaX11Platform platform, IWindowImpl? popupParent, bool ov
+                        return;
+                    }
+
+                    // The reason for using `_x11.Display` instead of `_x11.DeferredDisplay` is that XSHM requires pushing to a Display that can receive Events, in order to obtain the XShmCompletionEvent when rendering is complete.
```
DeferredDisplay should be used for rendering since compositor runs on a separate thread. You can get events from that connection too as previously discussed in https://github.com/AvaloniaUI/Avalonia/discussions/16690#discussioncomment-10378881 (see DeferredDisplayEvents class)

lindexi replied:
@kekekeks The DeferredDisplay do not call the `XSelectInput`, and I do not know how to recevice the XShmCompletionEvent from DeferredDisplay. Thank you.

kekekeks replied:
Separate XPeekEvent/XNextEvent (depending on queue exhaustion) loop when attempting to obtain next buffer would suffice. XSelectInput on render target creation.

kekekeks commented on src/Avalonia.X11/XShmExtensions/X11ShmDebugLogger.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace Avalonia.X11.XShmExtensions;
+
+internal static class X11ShmDebugLogger
```
`Logger.TryGet(LogEventLevel.Information, "X11")?.Log(...`

lindexi replied:
@kekekeks Thank you. I intend for the X11ShmDebugLogger to be used by Avalonia framework developers when investigating the XShm module, rather than by application developers using the Avalonia framework. Application developers should not be misled by the debug output from XShm. I'm not sure if this expectation is reasonable.

lindexi committed:
Pass the ShouldRenderOnUiThread to XShm

lindexi committed:
Fix thread safe issues, the GetOrCreateImage in X11ShmImageManager will be called in render thread and the OnXShmCompletion method will be called in UI thread

lindexi committed:
Re-define the XShmCompletionEvent

From https://www.x.org/releases/X11R7.5/doc/Xext/mit-shm.html

The XShmCompletionEvent event type value that will be used can be determined at run time

lindexi committed:
Fix compile fail in net standard

lindexi committed:
Add the MaxXShmSwapchainFrameCount property to X11ShmFramebufferContext

lindexi committed:
Add synchronously wait for current presentation count than max swapchain frame count

lindexi committed:
Update the XShm log

lindexi committed:
Make Analyzer happy

lindexi committed:
Bring the LastSize to the lock to fix the thread safe issues

lindexi committed:
Extract the ClearAvailableQueue in X11ShmImageManager

lindexi commented:
@kekekeks Can I ask any plan about this? The code I have now has some conflicts. If you plan to merge the code or do more advanced work, I will be motivated to solve the code conflicts.

kekekeks commented on src/Avalonia.X11/XShmExtensions/X11ShmImageManager.cs:
```diff
@@ -0,0 +1,124 @@
+            LastSize = size;
+        }
+
+        if (Context.ShouldRenderOnUiThread)
```
This is unbounded queue growth

kekekeks commented on src/Avalonia.X11/XShmExtensions/X11ShmImageManager.cs:
```diff
@@ -0,0 +1,124 @@
+        else if (_presentationCount > Context.MaxXShmSwapchainFrameCount)
+        {
+            // Specifically, allowing one additional frame beyond the maximum render limit is beneficial. This is because at any given moment, one frame might be in the process of being returned, and another might be currently rendering. Therefore, adding an extra frame in preparation for rendering can maximize rendering efficiency.
+            SpinWait.SpinUntil(() => _presentationCount <= Context.MaxXShmSwapchainFrameCount);
```
This is potentially 100% cpu usage by render thread when UI thread is busy.

kekekeks commented on src/Avalonia.X11/X11Window.cs:
```diff
@@ -665,21 +649,9 @@ private void OnEvent(ref XEvent ev)
-                _imeControl?.SetWindowActive(true);
-            }
-            else
+            else if ((int) ev.type == _x11ShmFramebufferSurface?.XShmCompletionType)
```
This should be polled by render thread before attempting to render. With synchronous wait if buffer queue is exhausted. Having this on UI thread causes issues described above.
